datafusion/optimizer/src/rewrite_cycle.rs

```diff
@@ -0,0 +1,412 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/// [`RewriteCycle`] API for executing a sequence of [TreeNodeRewriter]s in multiple passes.
+use std::ops::ControlFlow;
+
+use datafusion_common::{
+    tree_node::{Transformed, TreeNode, TreeNodeRewriter},
+    Result,
+};
+
+/// A builder with methods for executing a "rewrite cycle".
+///
+/// Often the results of one optimization rule can uncover more optimizations in other optimization
+/// rules. A sequence of optimization rules can be ran in multiple "passes" until there are no
+/// more optmizations to make.
+///
+/// The [RewriteCycle] handles logic for running these multi-pass loops.
+/// It applies a sequence of [TreeNodeRewriter]s to a [TreeNode] by calling
+/// [TreeNode::rewrite] in a loop - passing the output of one rewrite as the input to the next
+/// rewrite - until [RewriteCycle::max_cycles] is reached or until every [TreeNode::rewrite]
+/// returns a [Transformed::no] result in a consecutive sequence.
+#[derive(Debug)]
+pub struct RewriteCycle {
+    max_cycles: usize,
+}
+
+impl Default for RewriteCycle {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl RewriteCycle {
+    /// The default maximum number of completed cycles to run before terminating the rewrite loop.
+    /// You can override this default with [Self::with_max_cycles]
+    pub const DEFAULT_MAX_CYCLES: usize = 3;
+
+    /// Creates a new [RewriteCycle] with default options.
+    pub fn new() -> Self {
+        Self {
+            max_cycles: Self::DEFAULT_MAX_CYCLES,
+        }
+    }
+    /// Sets the [Self::max_cycles] to run before terminating the rewrite loop.
+    pub fn with_max_cycles(mut self, max_cycles: usize) -> Self {
+        self.max_cycles = max_cycles;
+        self
+    }
+
+    /// The maximum number of completed cycles to run before terminating the rewrite loop.
+    /// Defaults to [Self::DEFAULT_MAX_CYCLES].
+    pub fn max_cycles(&self) -> usize {
+        self.max_cycles
+    }
+
+    /// Runs a rewrite cycle on the given [TreeNode] using the given callback function to
```

Comment:
Could you please define how a "cycle" and "iteration" are related in this function? They are used in the apis below but I didn't see a definition of the terms and how they are related

I think it is that a cycle is composed of several iterations

Alternately, maybe "rewrites" is a better term than "iteration" as this is a "rewrite cycle" 🤔